### DISCUSSION TITLE
Representing TS index signatures in Zod

### DISCUSSION THREAD

**Discussion by binajmen** [colinhacks/zod#1200] [Q&A] [answered]
Hi,

Typescript describes index signatures as the following:
> Sometimes you don’t know all the names of a type’s properties ahead of time, but you do know the shape of the values.

I'm exactly in that case, having an object holding different translations:
```ts
{
  "en": "Hello!"
  "fr": "Salut!"
  ...
}
```

My current Zod solution is quite verbose:
```ts
z.object({
  en: z.string(),
  fr: z.string(),
  ... // <-- if I want to cover all the languages of the world, it will become quite verbose
}).partial().refine(val => val.en || val.fr || ...)
```

I also make the object partial with the only condition that at least one value exist.

Is there a Zod way to represent that? Something similar to `z.map(z.string(), z.string())` but for object would be awesome :)

**Comment by nullndr** [comment, marked as answer]
I would suggest you the use of [records](https://github.com/colinhacks/zod#records).

```ts
const schema = z.record(z.string(), z.string());
```

> I also make the object partial with the only condition that at least one value exist.

Add a `refine` that check for the length of keys in the object:

```ts
const schema = z.record(z.string(), z.string())
                .refine((val) => Object.keys(val).length > 0, {});
```

So

```ts
schema.parse({}); // Error
schema.parse({ en: "Hello!" }); // Ok
schema.parse({ en: "Hello!", fr: "Salut!" }); // Ok
```

**Reply by binajmen** [reply]
I could have sworn I read all the documentation! 😄 I guess I missed out that one, thank you 🙏 
```ts
const zI18n = z
  .record(z.string().length(2), z.string())
  .refine((val) => Object.keys(val).length > 0, {
    message: "At least one language required",
  });
```
